Sources/Amplitude/AMPIdentifyInterceptor.m

```diff
@@ -0,0 +1,285 @@
+//  all copies or substantial portions of the Software.
+//
+//  THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
+//  IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
+//  FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
+//  AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
+//  LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
+//  OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN
+//  THE SOFTWARE.
+//
+#ifndef AMPLITUDE_LOG_ERRORS
+#define AMPLITUDE_LOG_ERRORS 1
+#endif
+
+#ifndef AMPLITUDE_ERROR
+#if AMPLITUDE_LOG_ERRORS
+#   define AMPLITUDE_ERROR(fmt, ...) NSLog(fmt, ##__VA_ARGS__)
+#else
+#   define AMPLITUDE_ERROR(...)
+#endif
+#endif
+
+#import <Foundation/Foundation.h>
+#import "AMPConstants.h"
+#import "AMPEventUtils.h"
+#import "AMPIdentifyInterceptor.h"
+#import "AMPDatabaseHelper.h"
+
+@implementation AMPIdentifyInterceptor
+
+NSArray *_Nonnull _interceptOps;
+NSSet *_Nonnull _interceptOpsSet;
+BOOL _transferScheduled;
+NSOperationQueue *_Nonnull _backgroundQueue;
+AMPDatabaseHelper *_Nonnull _dbHelper;
+long _lastIdentifyInterceptorId;
+int _interceptedUploadPeriodSeconds;
+BOOL _disabled;
+
+-(id)initWithParams:(AMPDatabaseHelper *_Nonnull)dbHelper
+    backgroundQueue:(NSOperationQueue *_Nonnull)backgroundQueue
+{
+    if(self = [super init]) {
+        _dbHelper = dbHelper;
+        _backgroundQueue = backgroundQueue;
+        _lastIdentifyInterceptorId = -1;
+        _interceptedUploadPeriodSeconds = kAMPIdentifyUploadPeriodSeconds;
+        _transferScheduled = NO;
+        _disabled = NO;
+        _interceptOps = @[AMP_OP_SET]; // Notice: Supporting AMP_OP_SET_ONCE would require more complex merge logic
+        _interceptOpsSet = [NSSet setWithArray:_interceptOps];
+    }
+
+    return self;
+}
+
++ (instancetype _Nonnull)getIdentifyInterceptor:(AMPDatabaseHelper *_Nonnull) dbHelper
+                                backgroundQueue:(NSOperationQueue *_Nonnull)backgroundQueue
+{
+    return [[self alloc] initWithParams:dbHelper backgroundQueue:backgroundQueue];
+}
+
+// If this returns YES, the given Identify user property operations should be queued and batched later
+- (BOOL)hasInterceptOperationsOnly:(NSDictionary *_Nonnull)userPropertyOperations {
+    NSSet *operations = [NSSet setWithArray:[userPropertyOperations allKeys]];
+
+    return [operations isSubsetOfSet:_interceptOpsSet];
+}
+
+- (NSMutableDictionary *)intercept:(NSMutableDictionary *_Nonnull)event {
+    if (_disabled) {
+        return event;
+    }
+
+    NSString *eventType = [AMPEventUtils getEventType:event];
+
+    NSMutableDictionary *userPropertyOperations = [AMPEventUtils getUserProperties:event];
+    if (eventType == IDENTIFY_EVENT) {
+       // Check to intercept
+        if ([self hasInterceptOperationsOnly:userPropertyOperations] && [AMPEventUtils getGroups:event] == nil) {
```

Comment:
@qingzhuozhen I added events with `groups` since its a payed feature.

Reply:
OK, I don't feel strongly this should be exceptional. Though we could leave it like this for now.

Reply:
@izaaz do you have any opinions on this?

Reply:
Since the data gets sent within 30s to 1 minute, I don't think we need to skip the line or flush if groups are attached to identifies.

Reply:
Ok I am going to leave this but can remove it later if we want to further reduce volume.

Reply:
Will remove based on feedback from @izaaz